### ISSUE TITLE
[suggestion][tf.estimator] model with multiple labels

### ISSUE THREAD

**Issue by yanghoonkim** [tensorflow/tensorflow#17356]
How about let `labels` provided to `def model_fn(features, labels, mode, params):` be `dict` but not a tensor.
I'm now modeling a model with multiple outputs to be optimized, however, I can only use `features` to substitute `labels` because it only allow a tensor. I'm using version 1.4

**Comment by arthurarg** [comment]
I probably misunderstand your post but it seems that tf.estimator handles multi-dimension outputs (with labels taking in input a dict for multi-head models) - see tf.estimator.Estimator class definition:
 
Args:
      model_fn: Model function. Follows the signature:
        * Args:
          * `features`: This is the first item returned from the `input_fn`
                 passed to `train`, `evaluate`, and `predict`. This should be a
                 single `Tensor` or `dict` of same.
          * `labels`: This is the second item returned from the `input_fn`
                 passed to `train`, `evaluate`, and `predict`. This should be a
                 single `Tensor` or `dict` of same (for multi-head models). If
                 mode is `ModeKeys.PREDICT`, `labels=None` will be passed. If
                 the `model_fn`'s signature does not accept `mode`, the
                 `model_fn` must still be able to handle `labels=None`.
          * `mode`: Optional. Specifies if this training, evaluation or
                 prediction. See `ModeKeys`.
          * `params`: Optional `dict` of hyperparameters.  Will receive what
                 is passed to Estimator in `params` parameter. This allows
                 to configure Estimators from hyper parameter tuning.
          * `config`: Optional configuration object. Will receive what is passed
                 to Estimator in `config` parameter, or the default `config`.
                 Allows updating things in your model_fn based on configuration
                 such as `num_ps_replicas`, or `model_dir`.

**Comment by yanghoonkim** [comment]
Oh, I found that I only checked the following link:
https://www.tensorflow.org/versions/r1.4/extend/estimators
And in the link, `lables` in `model_fn` is only explained as:

> labels: A Tensor containing the labels passed to the model via input_fn. Will be empty for predict() calls, as these are the values the model will infer.

So did i make such a suggestion.

> * labels: This is the second item returned from the input_fn passed to train, evaluate, and predict. This should be a single Tensor or dict of same (for multi-head models). 

So this means that I can have multiple objective functions and different dimensions of label values? e.g.  the label of the first objective function with 5 dimension vector and the label of the second objective function with 2 dimension vector.

**Comment by arthurarg** [comment]
Yes, you can indeed. tf.estimator allows to handle it, then you can handle the multiple objectives however you want in your model_fn. See simple example bellow (taken from the TF website and modified for multi-dimension labels):
```import os
import urllib

import numpy as np
import tensorflow as tf


# Data sets
IRIS_TRAINING = "iris_training.csv"
IRIS_TRAINING_URL = "http://download.tensorflow.org/data/iris_training.csv"

IRIS_TEST = "iris_test.csv"
IRIS_TEST_URL = "http://download.tensorflow.org/data/iris_test.csv"

# If the training and test sets aren't stored locally, download them.
if not os.path.exists(IRIS_TRAINING):
  raw = urllib.urlopen(IRIS_TRAINING_URL).read()
  with open(IRIS_TRAINING, "w") as f:
    f.write(raw)

if not os.path.exists(IRIS_TEST):
  raw = urllib.urlopen(IRIS_TEST_URL).read()
  with open(IRIS_TEST, "w") as f:
    f.write(raw)

# Load datasets.
training_set = tf.contrib.learn.datasets.base.load_csv_with_header(
    filename=IRIS_TRAINING,
    target_dtype=np.int,
    features_dtype=np.float32)
test_set = tf.contrib.learn.datasets.base.load_csv_with_header(
    filename=IRIS_TEST,
    target_dtype=np.int,
    features_dtype=np.float32)

def build_estimator(
  model_dir,
  hidden_units,
  feature_columns,
  n_classes,
  output_dim):
  

  def _model_fn(features, labels, mode, params):
    """Model function for Estimator."""
    
    # One-hot encoding on targets
    print 'label shapes', [labels[l].shape for l in labels]
    
    labels = labels['y1']
    targets = tf.one_hot(labels, n_classes)

    # Connect the first hidden layer to input layer
    # (features["x"]) with relu activation
    first_hidden_layer = tf.layers.dense(features["x"], 10, activation=tf.nn.relu)
    # Connect the second hidden layer to first hidden layer with relu
    second_hidden_layer = tf.layers.dense(
        first_hidden_layer, 20, activation=tf.nn.relu)
    
    # Connect the third hidden layer to second hidden layer with relu
    third_hidden_layer = tf.layers.dense(
        second_hidden_layer, 20, activation=tf.nn.relu)

    # Connect the output layer to third hidden layer (no activation fn)
    output_layer = tf.layers.dense(third_hidden_layer, n_classes)

    # Reshape output layer to 1-dim Tensor to return predictions
    predictions = tf.argmax(output_layer, axis=1)

    # Provide an estimator spec for `ModeKeys.PREDICT`.
    if mode == tf.estimator.ModeKeys.PREDICT:
      return tf.estimator.EstimatorSpec(
          mode=mode,
          predictions={"predictions": predictions})

    # Calculate loss using mean squared error
    loss = tf.losses.softmax_cross_entropy(targets, output_layer)

    # Calculate root mean squared error as additional eval metric
    eval_metric_ops = {
        "accuracy": tf.metrics.accuracy(labels=labels, predictions=predictions)
    }
    
    optimizer = tf.train.GradientDescentOptimizer(
        learning_rate=0.1)
    train_op = optimizer.minimize(
        loss=loss, global_step=tf.train.get_global_step())

    # Provide an estimator spec for `ModeKeys.EVAL` and `ModeKeys.TRAIN` modes.
    return tf.estimator.EstimatorSpec(
        mode=mode,
        loss=loss,
        train_op=train_op,
        eval_metric_ops=eval_metric_ops)

  return tf.estimator.Estimator(
    model_fn=_model_fn)

# Specify that all features have real-value data
feature_columns = [tf.feature_column.numeric_column("x", shape=[4])]

# Build 3 layer DNN with 10, 20, 10 units respectively.
estimator = build_estimator# (#
classifier = estimator(feature_columns=feature_columns,
                                        hidden_units=[10, 20, 10],
                                        n_classes=3,
                                        output_dim=2,
                                        model_dir="/tmp/iris_model")
# Define the training inputs
train_input_fn = tf.estimator.inputs.numpy_input_fn(
    x={"x": np.array(training_set.data)},
    y={
        "y1": np.array(training_set.target),
        "y2": np.stack(
            [np.array(training_set.target) for i in xrange(2)],
            axis=1)
    },
    num_epochs=None,
    shuffle=True)

# Train model.
classifier.train(input_fn=train_input_fn, steps=2000)
```

**Comment by seabiscuit08** [comment]
> Yes, you can indeed. tf.estimator allows to handle it, then you can handle the multiple objectives however you want in your model_fn. See simple example bellow (taken from the TF website and modified for multi-dimension labels):
> 
> ```
> import urllib
> 
> import numpy as np
> import tensorflow as tf
> 
> 
> # Data sets
> IRIS_TRAINING = "iris_training.csv"
> IRIS_TRAINING_URL = "http://download.tensorflow.org/data/iris_training.csv"
> 
> IRIS_TEST = "iris_test.csv"
> IRIS_TEST_URL = "http://download.tensorflow.org/data/iris_test.csv"
> 
> # If the training and test sets aren't stored locally, download them.
> if not os.path.exists(IRIS_TRAINING):
>   raw = urllib.urlopen(IRIS_TRAINING_URL).read()
>   with open(IRIS_TRAINING, "w") as f:
>     f.write(raw)
> 
> if not os.path.exists(IRIS_TEST):
>   raw = urllib.urlopen(IRIS_TEST_URL).read()
>   with open(IRIS_TEST, "w") as f:
>     f.write(raw)
> 
> # Load datasets.
> training_set = tf.contrib.learn.datasets.base.load_csv_with_header(
>     filename=IRIS_TRAINING,
>     target_dtype=np.int,
>     features_dtype=np.float32)
> test_set = tf.contrib.learn.datasets.base.load_csv_with_header(
>     filename=IRIS_TEST,
>     target_dtype=np.int,
>     features_dtype=np.float32)
> 
> def build_estimator(
>   model_dir,
>   hidden_units,
>   feature_columns,
>   n_classes,
>   output_dim):
>   
> 
>   def _model_fn(features, labels, mode, params):
>     """Model function for Estimator."""
>     
>     # One-hot encoding on targets
>     print 'label shapes', [labels[l].shape for l in labels]
>     
>     labels = labels['y1']
>     targets = tf.one_hot(labels, n_classes)
> 
>     # Connect the first hidden layer to input layer
>     # (features["x"]) with relu activation
>     first_hidden_layer = tf.layers.dense(features["x"], 10, activation=tf.nn.relu)
>     # Connect the second hidden layer to first hidden layer with relu
>     second_hidden_layer = tf.layers.dense(
>         first_hidden_layer, 20, activation=tf.nn.relu)
>     
>     # Connect the third hidden layer to second hidden layer with relu
>     third_hidden_layer = tf.layers.dense(
>         second_hidden_layer, 20, activation=tf.nn.relu)
> 
>     # Connect the output layer to third hidden layer (no activation fn)
>     output_layer = tf.layers.dense(third_hidden_layer, n_classes)
> 
>     # Reshape output layer to 1-dim Tensor to return predictions
>     predictions = tf.argmax(output_layer, axis=1)
> 
>     # Provide an estimator spec for `ModeKeys.PREDICT`.
>     if mode == tf.estimator.ModeKeys.PREDICT:
>       return tf.estimator.EstimatorSpec(
>           mode=mode,
>           predictions={"predictions": predictions})
> 
>     # Calculate loss using mean squared error
>     loss = tf.losses.softmax_cross_entropy(targets, output_layer)
> 
>     # Calculate root mean squared error as additional eval metric
>     eval_metric_ops = {
>         "accuracy": tf.metrics.accuracy(labels=labels, predictions=predictions)
>     }
>     
>     optimizer = tf.train.GradientDescentOptimizer(
>         learning_rate=0.1)
>     train_op = optimizer.minimize(
>         loss=loss, global_step=tf.train.get_global_step())
> 
>     # Provide an estimator spec for `ModeKeys.EVAL` and `ModeKeys.TRAIN` modes.
>     return tf.estimator.EstimatorSpec(
>         mode=mode,
>         loss=loss,
>         train_op=train_op,
>         eval_metric_ops=eval_metric_ops)
> 
>   return tf.estimator.Estimator(
>     model_fn=_model_fn)
> 
> # Specify that all features have real-value data
> feature_columns = [tf.feature_column.numeric_column("x", shape=[4])]
> 
> # Build 3 layer DNN with 10, 20, 10 units respectively.
> estimator = build_estimator# (#
> classifier = estimator(feature_columns=feature_columns,
>                                         hidden_units=[10, 20, 10],
>                                         n_classes=3,
>                                         output_dim=2,
>                                         model_dir="/tmp/iris_model")
> # Define the training inputs
> train_input_fn = tf.estimator.inputs.numpy_input_fn(
>     x={"x": np.array(training_set.data)},
>     y={
>         "y1": np.array(training_set.target),
>         "y2": np.stack(
>             [np.array(training_set.target) for i in xrange(2)],
>             axis=1)
>     },
>     num_epochs=None,
>     shuffle=True)
> 
> # Train model.
> classifier.train(input_fn=train_input_fn, steps=2000)
> ```

Hello,I see this is for training label for y1,but how can it train y2?Since the EstimatorSpec takes only one loss argument?

**Comment by sachinprasadhs** [comment]
Tf.estimator.Estimator uses Tensorflow 1 style of coding and is not recommended for Tensorflow 2.x. There will not be any new addition to the estimators. 
You can follow. the detailed explanation [here](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator) and close this issue. Thanks!
